exercises/tutorial_halfday/ex8_matrix-transpose-local-array.cpp

```diff
@@ -0,0 +1,348 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include <cmath>
+#include <cstdlib>
+#include <cstring>
+#include <iostream>
+
+#include "RAJA/RAJA.hpp"
+#include "memoryManager.hpp"
+
+/*
+ *  Exercise #8: Matrix Transpose with Local Array
+ *
+ *  In this exercise, your program will carry out the
+ *  transpose of a matrix A using a tiling algorithm and a RAJA local array.
+ *  Unlike the previous exercise, each tile will be stored within
+ *  a RAJA local array. As part of the exercise you will have to provide
+ *  the transpose as a second matrix At.
+ *
+ *  This file contains a C-style variant of the algorithm as well as the
+ *  RAJA kernel for a RAJA variant. You will have to
+ *  implement the RAJA policy for the sequential, and OpenMP variants.
+ *  If you have access to a GPU and a CUDA compiler, try using constructing
+ *  the CUDA policy.
+ *
+ *  RAJA features shown:
+ *    - Basic usage of 'RAJA::kernel' abstractions for nested loops
+ *       - Multiple lambdas
+ *       - Tile statement
+ *       - ForICount statement
+ *       - RAJA local arrays
+ *
+ * If CUDA is enabled, CUDA unified memory is used.
+ */
+
+//
+// Define dimensionality of matrices
+//
+const int DIM = 2;
+
+//
+// Function for checking results
+//
+template <typename T>
+void checkResult(RAJA::View<T, RAJA::Layout<DIM>> Atview, int N_r, int N_c);
+
+//
+// Function for printing results
+//
+template <typename T>
+void printResult(RAJA::View<T, RAJA::Layout<DIM>> Atview, int N_r, int N_c);
+
+
+int main(int RAJA_UNUSED_ARG(argc), char **RAJA_UNUSED_ARG(argv[]))
+{
+
+  std::cout << "\n\nRAJA shared matrix transpose example...\n";
+
+  //
+  // Define num rows/cols in matrix
+  //
+  const int N_r = 267;
+  const int N_c = 251;
+
+  //
+  // Allocate matrix data
+  //
+  int *A = memoryManager::allocate<int>(N_r * N_c);
+  int *At = memoryManager::allocate<int>(N_r * N_c);
+
+  //
+  // In the following implementations of matrix transpose, we
+  // use RAJA 'View' objects to access the matrix data. A RAJA view
+  // holds a pointer to a data array and enables multi-dimensional indexing
+  // into the data.
+  //
+  RAJA::View<int, RAJA::Layout<DIM>> Aview(A, N_r, N_c);
+  RAJA::View<int, RAJA::Layout<DIM>> Atview(At, N_c, N_r);
```

Comment:
Similar comment about permuting the view for At so both reads and writes are stride-1.